File: src/algorithms/trees/bst-operations/bst-to-greater-tree-iterative/sources/BSTToGreaterTreeIterative_test.java
// javac *.java && java -ea BSTToGreaterTreeIterative_test
public class BSTToGreaterTreeIterative_test {
    static BSTNode makeNode(int value, BSTNode left, BSTNode right) {
        BSTNode node = new BSTNode(value);
        node.left = left;
        node.right = right;
        return node;
    }

    static BSTNode leaf(int value) { return new BSTNode(value); }

    public static void main(String[] args) {
        BSTToGreaterTreeIterative bgti = new BSTToGreaterTreeIterative();

        // test: transforms 3-node BST
        BSTNode tree1 = makeNode(2, leaf(1), leaf(3));
        BSTNode result1 = bgti.bstToGreaterTreeIterative(tree1);
        assert result1.value == 5 : "Root should be 5, got " + result1.value;
        assert result1.right.value == 3 : "Right should be 3";
        assert result1.left.value == 6 : "Left should be 6";

        // test: single node
        BSTNode result2 = bgti.bstToGreaterTreeIterative(leaf(7));
        assert result2.value == 7 : "Single node should stay 7";

        // test: null tree
        assert bgti.bstToGreaterTreeIterative(null) == null : "Null tree should return null";

        System.out.println("All tests passed!");
    }
}

File: src/algorithms/trees/bst-operations/bst-to-greater-tree-iterative/__tests__/bst-to-greater-tree-iterative_test.go
package main

import "testing"

func makeGTIterNode(value int, left *BSTNode, right *BSTNode) *BSTNode {
	return &BSTNode{value: value, left: left, right: right}
}

func gtiLeaf(value int) *BSTNode {
	return &BSTNode{value: value}
}

func TestBSTToGreaterTreeIter3NodeBST(t *testing.T) {
	// Node 3->3, node 2->5, node 1->6
	tree := makeGTIterNode(2, gtiLeaf(1), gtiLeaf(3))
	result := bstToGreaterTreeIterative(tree)
	if result == nil || result.value != 5 {
		t.Errorf("root should be 5, got %v", result)
	}
	if result.right == nil || result.right.value != 3 {
		t.Error("right should be 3")
	}
	if result.left == nil || result.left.value != 6 {
		t.Error("left should be 6")
	}
}

func TestBSTToGreaterTreeIterSingleNode(t *testing.T) {
	result := bstToGreaterTreeIterative(gtiLeaf(7))
	if result == nil || result.value != 7 {
		t.Error("single node should remain 7")
	}
}

func TestBSTToGreaterTreeIterNilTree(t *testing.T) {
	if bstToGreaterTreeIterative(nil) != nil {
		t.Error("nil tree should return nil")
	}
}
